conanfile.py

```diff
@@ -0,0 +1,82 @@
+from conans import CMake, ConanFile, tools
```

Comment:
This won't help at all for getting seasocks in cci, because it doesn't follow cci practices.
Also, in my own experience, using a conanfile makes it easier to test on ci.
If you ever lose your mind and want to add support for Windows, then you can use it to build on appveyor.
A simple `conan install.. && conan build .. & conan test ..` *should work*™ .

In theory, now you can add a job on travis that uses this conanfile.
The `test()` method can be used to run the unit tests.
I haven't added a `test_package`, because the intention of this recipe is not to package `seasocks`, but to allow it to build (and test).

Reply:
I see. I can't see that we'd ever use conan to build seasocks in this case (it has so few upstream dependencies). My projects that _do_ have upstream deps do use conan, but I don't want to rely on it here. However, that doesn't mean this isn't super useful!

We will definitely add a conan test job once this is working, how would you suggest doing so with your changes here?

> conan install.. && conan build .. & conan test ..

Aside: I do so hate the multi-stage aspect of conan. In all my experience, if there are manual steps, then people will forget them! In my own projects I always use the conan.cmake file to make conan part of the build itself...I've spent too long debugging things like "oh, you forgot to conan install:? You're probably building against the old libraries...argh!")

Reply:
> We will definitely add a conan test job once this is working, how would you suggest doing so with your changes here?

This requires adding new jobs, that will run the conan commands.
I would add a `CONAN_INSTALL_ARGS` environment variable to `.travis.yml`.


> Aside: I do so hate the multi-stage aspect of conan. In all my experience, if there are manual steps, then people will forget them! In my own projects I always use the conan.cmake file to make conan part of the build itself...I've spent too long debugging things like "oh, you forgot to conan install:? You're probably building against the old libraries...argh!")

It's a matter of taste I guess.
I always try to make my project package manager agnostic.
This allows me to build with the libraries of a linux package manager.

Reply:
Oops! I meant to reply!!! Not edit your comment :blush:

I tried to say:

This requires adding new jobs, that will run the conan commands.
I would add a CONAN_INSTALL_ARGS environment variable to .travis.yml.

I get that, just quickly, how might one actually build seasocks using conan? conan install .. && conan build .. && conan test .. like you mentioned above? Would that actually run the tests? Thanks again for teaching me this.

I always try to make my project package manager agnostic.
Got it: that makes sense. In my world having the one way that just works and is reproducible is most important, but I have the luxury of mostly getting to choose this within my team etc

Reply:
> I get that, just quickly, how might one actually build seasocks using conan? conan install .. && conan build .. && conan test .. like you mentioned above? Would that actually run the tests? Thanks again for teaching me this.

I was wrong there. The intended use of `conan test` is only to test conan packages.
What can be done is use the [virtualenv generator](https://docs.conan.io/en/latest/reference/generators/virtualenv.html). The tests can then be run from any script (bash/python/ctest/...). The virtualenv makes sure that paths of all dependencies are in the search path (LD_LIBRARY_PATH/PATH/DYLD_LIBRARY_PATH/...). This is especially required for shared library dependencies.

> I always try to make my project package manager agnostic.
> Got it: that makes sense. In my world having the one way that just works and is reproducible is most important, but I have the luxury of mostly getting to choose this within my team etc

Check the `conanfile.py` of thie recipe. It writes a wrapper `CMakeLists.txt` in the build directory that includes the source directory. It is more or less the inverse approach as your `conan.cmake`. (conan.cmake's driver is cmake. In this pr, the driver is conan)

Reply:
I'll try and work through that once we're merged in. I'm not quite following but I imagine trying to do it will elucidate more than reading dry documentation.